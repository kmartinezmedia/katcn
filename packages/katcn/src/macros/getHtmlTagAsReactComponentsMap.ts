import { SyntaxKind } from 'ts-morph';
import type { PrimitiveType, ValidHtmlTag } from '../types';
import { katcnComponentsSourceFile } from './tsMorph';
import { typesToConstants } from './typesToConstants';

type HtmlAsComponent = {
  name: string;
  defaultProps: Record<string, PrimitiveType>;
};

const validHtmlTags = typesToConstants<ValidHtmlTag>('ValidHtmlTag');

function isValidHtmlTag(tag: string | ValidHtmlTag): tag is ValidHtmlTag {
  return validHtmlTags.includes(tag as ValidHtmlTag);
}

export function getHtmlTagAsReactComponentsMap() {
  const htmlAsComponentsMap: { [key in ValidHtmlTag]?: HtmlAsComponent } = {};

  const components = katcnComponentsSourceFile.getDescendantsOfKind(
    SyntaxKind.FunctionDeclaration,
  );

  for (const component of components) {
    const name = component.getName();
    if (!name) continue;
    const jsdocs = component.getJsDocs();
    const htmlJsDoc = jsdocs
      .flatMap((doc) => doc.getTags())
      .find((item) => item.getTagName() === 'html')
      ?.getText()
      .replace('@html ', '')
      .replaceAll(/\s/g, '')
      .trim();

    const htmlTags = htmlJsDoc?.split(',') ?? [];
    if (htmlTags.length > 1) {
      for (const htmlTag of htmlTags) {
        if (isValidHtmlTag(htmlTag)) {
          htmlAsComponentsMap[htmlTag] = {
            name,
            defaultProps: { as: htmlTag },
          };
        }
      }
    } else {
      const htmlTag = htmlTags[0];
      if (isValidHtmlTag(htmlTag)) {
        htmlAsComponentsMap[htmlTag] = {
          name,
          defaultProps: {},
        };
      }
    }
  }

  /** For some reason bun macros require this to properly output the object */
  return htmlAsComponentsMap;
}
