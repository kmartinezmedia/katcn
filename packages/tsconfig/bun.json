{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Bun",
  "extends": "./node.json",
  "compilerOptions": {
    "types": ["bun-types"]
  }
}
