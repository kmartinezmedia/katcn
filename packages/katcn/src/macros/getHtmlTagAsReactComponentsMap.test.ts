/// <reference types="bun-types" />

import { describe, expect, it } from 'bun:test';
import { getHtmlTagAsReactComponentsMap } from './getHtmlTagAsReactComponentsMap';

describe('getHtmlTagAsReactComponentsMap', () => {
  it('Matches snapshot', () => {
    const value = getHtmlTagAsReactComponentsMap();
    expect(value).toMatchSnapshot();
  });

  it('Should return a map of html components to react components', () => {
    const htmlAsComponentsMap = getHtmlTagAsReactComponentsMap();
    expect(htmlAsComponentsMap).toBeDefined();
    expect(htmlAsComponentsMap.div).toBeDefined();
    expect(htmlAsComponentsMap.p).toMatchObject({
      name: 'Text',
      defaultProps: { as: 'p' },
    });
  });
});
