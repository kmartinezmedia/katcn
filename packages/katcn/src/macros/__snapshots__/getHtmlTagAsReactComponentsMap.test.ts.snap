// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getHtmlTagAsReactComponentsMap Matches snapshot 1`] = `
{
  "a": {
    "defaultProps": {},
    "name": "Anchor",
  },
  "article": {
    "defaultProps": {},
    "name": "Article",
  },
  "aside": {
    "defaultProps": {},
    "name": "Aside",
  },
  "button": {
    "defaultProps": {},
    "name": "Pressable",
  },
  "div": {
    "defaultProps": {},
    "name": "Box",
  },
  "em": {
    "defaultProps": {
      "as": "em",
    },
    "name": "Text",
  },
  "footer": {
    "defaultProps": {},
    "name": "Footer",
  },
  "h1": {
    "defaultProps": {
      "as": "h1",
    },
    "name": "Text",
  },
  "h2": {
    "defaultProps": {
      "as": "h2",
    },
    "name": "Text",
  },
  "h3": {
    "defaultProps": {
      "as": "h3",
    },
    "name": "Text",
  },
  "h4": {
    "defaultProps": {
      "as": "h4",
    },
    "name": "Text",
  },
  "h5": {
    "defaultProps": {
      "as": "h5",
    },
    "name": "Text",
  },
  "h6": {
    "defaultProps": {
      "as": "h6",
    },
    "name": "Text",
  },
  "header": {
    "defaultProps": {},
    "name": "Header",
  },
  "i": {
    "defaultProps": {
      "as": "i",
    },
    "name": "Text",
  },
  "img": {
    "defaultProps": {},
    "name": "Image",
  },
  "input": {
    "defaultProps": {},
    "name": "TextInput",
  },
  "label": {
    "defaultProps": {
      "as": "label",
    },
    "name": "Text",
  },
  "li": {
    "defaultProps": {},
    "name": "ListItem",
  },
  "main": {
    "defaultProps": {},
    "name": "Main",
  },
  "nav": {
    "defaultProps": {},
    "name": "Nav",
  },
  "ol": {
    "defaultProps": {
      "as": "ol",
    },
    "name": "List",
  },
  "option": {
    "defaultProps": {},
    "name": "Option",
  },
  "p": {
    "defaultProps": {
      "as": "p",
    },
    "name": "Text",
  },
  "pre": {
    "defaultProps": {
      "as": "pre",
    },
    "name": "Text",
  },
  "s": {
    "defaultProps": {
      "as": "s",
    },
    "name": "Text",
  },
  "section": {
    "defaultProps": {},
    "name": "Section",
  },
  "select": {
    "defaultProps": {},
    "name": "Select",
  },
  "span": {
    "defaultProps": {
      "as": "span",
    },
    "name": "Text",
  },
  "strong": {
    "defaultProps": {
      "as": "strong",
    },
    "name": "Text",
  },
  "sub": {
    "defaultProps": {
      "as": "sub",
    },
    "name": "Text",
  },
  "sup": {
    "defaultProps": {
      "as": "sup",
    },
    "name": "Text",
  },
  "table": {
    "defaultProps": {},
    "name": "Table",
  },
  "tbody": {
    "defaultProps": {},
    "name": "TableBody",
  },
  "td": {
    "defaultProps": {},
    "name": "TableCell",
  },
  "textarea": {
    "defaultProps": {},
    "name": "TextArea",
  },
  "th": {
    "defaultProps": {},
    "name": "TableHeadCell",
  },
  "thead": {
    "defaultProps": {},
    "name": "TableHead",
  },
  "tr": {
    "defaultProps": {},
    "name": "TableRow",
  },
  "u": {
    "defaultProps": {
      "as": "u",
    },
    "name": "Text",
  },
  "ul": {
    "defaultProps": {
      "as": "ul",
    },
    "name": "List",
  },
}
`;
